pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -1028,8 +1031,21 @@ func isGatewayMatch(gateway string, gatewayNames []string) bool {
 	return false
 }
 
-func buildGatewayVirtualHostDomains(hostname string, port int) []string {
+func buildGatewayVirtualHostDomains(service *model.Service, hostname string, port int) []string {
 	domains := []string{hostname}
+	if service != nil && service.Attributes.ServiceRegistry == provider.Kubernetes {
+		domains = append(domains,
+			service.Attributes.Name,
```

Comment:
This is dangerous. Consider VS in `bar` namespace AND namespace in `foo` namespace for `helloworld`.

1. We may get duplicate domains. I think we filter them out, but will give a ton of warnings + behavior based on creation time of service which is not great.
2. Say the `foo` namepace wins the conflict. If the `bar` namespace calls `helloworld`, they will unexpectedly go to `foo`.

Knative does something very similar with their cluster-local-gateway, and they do NOT support this form.

Reply:
Yes, for different namespaces, when there are two same name service.

Reply:
Right.. its not safe.

If I have `helloworld` in a sidecar, its fine because we know the namespace of the client (its in the same pod!). For a gateway, we have clients from all namespaces.